### PR TITLE
librbd: get_parent API method should properly handle migrating image

### PR DESCRIPTION
**Pull request by dillaman**
The true parent of a migrating parent is actually the parent of the
migration source image. There are other API methods available to
retreive the details of the migration source.

Fixes: http://tracker.ceph.com/issues/37998
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/librbd/api/Image.cc [hunk 1]
```diff
@@ -168,33 +169,40 @@ int Image<I>::get_parent(I *ictx,
...
-  if (ictx->snap_id == CEPH_NOSNAP) {
-    parent_spec = ictx->parent_md.spec;
-  } else {
-    r = ictx->get_parent_spec(ictx->snap_id, &parent_spec);
-    if (r < 0) {
-      lderr(cct) << "error looking up snapshot id " << ictx->snap_id << dendl;
-      return r;
+  auto parent = ictx->parent;
+  bool release_parent_locks = false;
+  BOOST_SCOPE_EXIT_ALL(ictx, &parent, &release_parent_locks) {
```
Comment (trociny) [inline src/librbd/api/Image.cc]: nit: it looks like captured `parent` is not used here.

File: src/librbd/api/Image.cc [hunk 2]
```diff
@@ -168,33 +169,40 @@ int Image<I>::get_parent(I *ictx,
...
+      ictx->parent->snap_lock.put_read();
     }
-    if (parent_spec.pool_id == -1) {
-      return -ENOENT;
+  };
+
+  // if a migration is in-progress, the true parent is the parent
+  // of the migration source image
+  if (!ictx->migration_info.empty() && parent != nullptr) {
+    parent = parent->parent;
```
Comment (trociny) [inline src/librbd/api/Image.cc]: nit: aren't we allowed to do this only after acquiring `ictx->parent->parent_lock`?

File: src/librbd/api/Image.cc [hunk 3]
```diff
@@ -168,33 +169,40 @@ int Image<I>::get_parent(I *ictx,
...
+      ictx->parent->snap_lock.put_read();
     }
-    if (parent_spec.pool_id == -1) {
-      return -ENOENT;
+  };
+
+  // if a migration is in-progress, the true parent is the parent
+  // of the migration source image
+  if (!ictx->migration_info.empty() && parent != nullptr) {
+    if (parent != nullptr) {
```
Comment (trociny) [inline src/librbd/api/Image.cc]: nit: no need to check  for `nullptr` here as it is already checked one line above.

File: src/librbd/api/Image.cc (hunk 1 above)
Comment (trociny) [reply src/librbd/api/Image.cc]: This version also contained the captured `parent` which seems to be unused?

**Comment by dillaman** [comment]
@trociny updated

**Comment by trociny** [comment]
@dillaman `rbd info` needs updating:
```
% rbd status clone1
Watchers: none
Migration:
        source: rbd/clone (1be164a83ddf3)
        destination: rbd/clone1 (1be269a4b30bb)
        state: prepared
% rbd info clone1                   
rbd image 'clone1':
        size 1 GiB in 256 objects
        order 22 (4 MiB objects)
        snapshot_count: 0
        id: 1be269a4b30bb
        block_name_prefix: rbd_data.1be269a4b30bb
        format: 2
        features: layering, exclusive-lock, object-map, fast-diff, deep-flatten, operations
        op_features: clone-child
        flags: 
        create_timestamp: Tue Feb 12 10:45:23 2019
        access_timestamp: Tue Feb 12 10:45:23 2019
        modify_timestamp: Tue Feb 12 10:45:23 2019
        parent: rbd/test@snap1 (migration source)
        overlap: 1 GiB
```
Note `(migration source)` in the `parent` field, which is not valid now.

**Comment by dillaman** [comment]
@trociny Fixed -- and also fixed the missing mapping for the migration feature bitmask.

**Comment by trociny** [comment]
@dillaman Jenkins got stuck on librbd unittests [1]. Unfortunately one can't see what exactly test it was and might be not related to this PR. Trying again.

@ceph-jenkins retest this please

[1] https://jenkins.ceph.com/job/ceph-pull-requests/13787/console

**Pull request merged**